src/include/common/types/value.h

```diff
@@ -215,10 +203,22 @@ class Value {
         internalID_t internalIDVal;
     } val;
     std::string strVal;
-    std::vector<std::unique_ptr<Value>> nestedTypeVal;
-    // TODO(Ziyi): remove these two fields once we implemented node/rel using struct.
-    std::unique_ptr<NodeVal> nodeVal;
-    std::unique_ptr<RelVal> relVal;
+
+private:
+    LogicalType dataType;
```

Comment:
I'm not sure how much the `LogicalType` copying is factoring into the iteration costs in comparison to the child values themselves, but since the `LogicalType` of the child values is included in the parent value's `LogicalType.extraTypeInfo`, maybe both the `ExtraTypeInfo` subclasses and `Value` should used `shared_ptr` to store the `LogicalType`, and the pointer could just be copied for the child types to refer to the type in the parent. It might be more suited to a follow-up PR since it's mostly distinct from these changes.

It might only be significant for highly nested types though, and avoiding re-creating the values also has the side effect of avoiding re-creating their LogicalTypes, so these changes should already provide most of the benefits.

Reply:
Move the LogicalType refactor as a separate PR (#1852). Though I choose the unique_ptr solution rather than the raw pointer one.